### PR TITLE
Add source code for Chapter 13

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+SharedBox Truckload::operator[](size_t index) const
+{
+  size_t count{};             // Package count
+  for (Package* package{ m_head }; package; package = package->m_next)
+  {
+    if (count++ == index)      // Up to index yet?
+      return package->m_box;   // If so return the pointer to Box
+  }
+  return nullptr;
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp:118]: This should be:
`return nullBox;`

Comment (MarcGregoire) [reply Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp:118]: And somewhere you need:
`static SharedBox nullBox {};   // Pointer to nullptr`

File: Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }                                
+                                     // Move both pointers along (mind the order!)
+    previous = current;              //  - first current becomes the new previous
+    current = current->m_next;       //  - then move current along to the next Package
+  }
+
+  return false;     // Return false: boxToRemove was not found
+}
+
+SharedBox Truckload::operator[](size_t index) const
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp:110]: This should return a reference according to the text

File: Examples/NoModules/Chapter 13/Ex13_11/Truckload.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  ~Truckload();                     // Destructor
+
+  class Iterator;      // Declaration of a public nested class, Truckload::Iterator
+
+  Iterator getIterator() const;
+
+  void addBox(SharedBox box);       // Add a new SharedBox
+  bool removeBox(SharedBox box);    // Remove a Box from the Truckload
+
+  SharedBox operator[](size_t index) const;   // Overloaded subscript operator
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 13/Ex13_11/Truckload.h:30]: This should return a reference according to the text

File: Examples/NoModules/Chapter 13/Ex13_12/Message.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <cstring> // For std::strlen() and std::strcpy()
+
+class Message
+{
+public:
+  explicit Message(const char* text = "")
+    : m_text(new char[std::strlen(text) + 1]) // Caution: include the null character!
+  {
+    std::strcpy(m_text, text);        // Mind the order: strcpy(destination, source)!
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 13/Ex13_12/Message.h:12]: Just saying, but VC++ gives error:
`'strcpy': This function or variable may be unsafe. Consider using strcpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.`

File: Examples/NoModules/Chapter 13/Ex13_12B/Ex13_12B.cpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// Sneak preview: use copy-and-swap for copy assignment operator
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 13/Ex13_12B/Ex13_12B.cpp:1]: My draft text does not have an Ex13_12B. Maybe you added it in the final text?

File: Exercises/NoModules/Chapter 13/Soln13_01/Soln13_01.cpp [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+// to post-multiply by an integer
+#include <iostream>
+#include "Box.h"
+
+int main()
+{
+  Box box {2, 3, 4};
+  std::cout << "Box is " << box << std::endl;
+  unsigned n {3};
+  Box newBox = box * n;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_01/Soln13_01.cpp:12]: uniform init

File: Exercises/NoModules/Chapter 13/Soln13_02/Soln13_02.cpp [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+// Implementing the * operator for the Box class to pre-multiply by a number
+#include <iostream>
+#include "Box.h"
+
+int main()
+{
+  Box box {2, 3, 4};
+  std::cout << "Box is " << box << std::endl;
+  unsigned n {3};
+  Box newBox = n * box;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_02/Soln13_02.cpp:11]: uniform init

File: Exercises/NoModules/Chapter 13/Soln13_03/Box.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+{
+  m_length *= factor;
+  m_width  *= factor;
+  m_height *= factor;
+  return *this;   // Return a reference to the left operand (convention)
+}
+
+Box& Box::operator/=(double divisor)
+{
+  m_length /= divisor;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_03/Box.cpp:51]: check for division by zero?

File: Exercises/NoModules/Chapter 13/Soln13_05/Box.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  }
+  std::partial_ordering operator<=>(double otherVolume) const 
+  { 
+    return volume() <=> otherVolume;
+  }
+
+  bool operator==(const Box& otherBox) const = default;
+
+  // Unary negation operator (!box is true if the Box has no volume)
+  bool operator!() const { return volume() == 0; }
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_05/Box.h:34]: This operator is not needed.
Just implementing the bool() conversion op is enough for !box to work.

Comment (MarcGregoire) [reply Exercises/NoModules/Chapter 13/Soln13_05/Box.h:34]: ok, that seemed to be the goal of this exercises, after reading the next exercise...

File: Exercises/NoModules/Chapter 13/Soln13_06/Soln13_06.cpp [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  testBox(box1);
+  
+  std::cout << std::endl;;
+
+  Box box2{0, 0, 0};
+  std::cout << "box2 is " << box2 << std::endl;
+  testBox(box2);  
+
+  // bool b1 = box1;    /* Does not compile! */
+  bool b2 = static_cast<bool>(box2);  // Needs an explicit type conversion
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_06/Soln13_06.cpp:38]: uniform init

File: Exercises/NoModules/Chapter 13/Soln13_07/Rational.h [hunk 11]
```diff
@@ -0,0 +1,126 @@
+// Rational.h
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_07/Rational.h:1]: I did not verify the math formulas. I'm assuming they are correct...

File: Exercises/NoModules/Chapter 13/Soln13_08/PRNG.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+public:
+  PseudoRandomNumberGenerator(int n = 0)
+    : m_n{ n }
+  {}
+
+  // A function call operator (no parameters, return type int)
+  // Unlike many function call operators, this one alters the state of the object.
+  int operator()()
+  {
+    const int current = m_n;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_08/PRNG.h:15]: uniform init

File: Exercises/NoModules/Chapter 13/Soln13_09/Truckload.cpp [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+Truckload& Truckload::operator=(const Truckload& other)
+{
+  if (&other != this)   // Do not forget: avoid issues with self-assignment!
+  {
+    delete m_head;              // Delete all current packages
+    m_head = m_tail = nullptr;  // Reset both pointers
+
+    // Same as copy constructor 
+    // (see Chapter 17 for the copy-and-swap idiom that allows you to avoid
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_09/Truckload.cpp:50]: It seems you added something to this chapter for copy-swap.
At least I see an Ex13_12B (which is not yet in my draft though) that explains copy-swap for Message.
If that's correct, then you might as well use copy-swap in this solution as well.

File: Exercises/NoModules/Chapter 13/Soln13_09/Soln13_09.cpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  std::random_device seeder;        // True random number generator to obtain a seed (slow)
+  std::default_random_engine generator{ seeder() };    // Efficient pseudo-random generator
+  std::uniform_real_distribution distribution{ 1.0, max }; // Generate in [1, max) interval
+  return std::bind(distribution, generator);           //... and in the darkness bind them!
+}
+
+int main()
+{
+  const double limit{ 99.0 };    // Upper limit on Box dimensions
+  auto random = createUniformPseudoRandomNumberGenerator(limit);
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_09/Soln13_09.cpp:21]: uniform init

File: Exercises/NoModules/Chapter 13/Soln13_01/Soln13_01.cpp [hunk 15]
```diff
@@ -0,0 +1,14 @@
+// Exercise 12-1
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 13/Soln13_01/Soln13_01.cpp:1]: 13-1

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Examples/NoModules/Chapter 13/Ex13_07/Integer.h [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+{
+  return one.getValue() - other.getValue();
+}
+inline Integer operator*(const Integer& one, const Integer& other)
+{
+  return one.getValue() * other.getValue();
+}
+inline Integer operator/(const Integer& one, const Integer& other)
+{
+  return one.getValue() / other.getValue();
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_07/Integer.h:28]: And do what? Exceptions are not known yet. Nah, this Integer class can just do what regular ints do upon division by zero.

File: Examples/NoModules/Chapter 13/Ex13_09/Box.h [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+{
+public:
+  // Constructors
+  Box() = default;
+  Box(double l, double w, double h) : m_length{ l }, m_width{ w }, m_height{ h } {}
+
+  Box& operator++();         // Prefix ++operator
+  const Box operator++(int); // Postfix operator++
+  Box& operator--();         // Prefix --operator
+  const Box operator--(int); // Postfix operator--
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_09/Box.h:18]: The book explains this: that is apparently the correct way of doing things (Horton found that).

File: Examples/NoModules/Chapter 13/Ex13_09/Ex13_09.cpp [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  std::cout << "Postfix increment evaluates to the original object: "
+    << theBox++ << std::endl;
+
+  std::cout << "After postfix increment: " << theBox << std::endl;
+
+  std::cout << "Prefix decrement evaluates to the decremented object: "
+    << --theBox << std::endl;
+  std::cout << "After prefix decrement: " << theBox << std::endl;
+}
+
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_09/Ex13_09.cpp:20]: The text has been updated for this, yes.

File: Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp (hunk 1 above)
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp:118]: I actually need `static inline SharedBox nullBox {}; // Pointer to nullptr`. Without the `inline` this of course does not work. One more for the errata...

File: Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp (hunk 2 above)
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_11/Truckload.cpp:110]: Naturally, that's the point of the entire example...

File: Examples/NoModules/Chapter 13/Ex13_12/Message.h (hunk 4 above)
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_12/Message.h:12]: Of course it does. Yet afaik MSVC is the only compiler that supports `strcpy_s` (and these are not even standard compliant: see https://stackoverflow.com/a/57915261), so I'm not about to use that. In the text I stress plenty of times not to use the C functions, and use std::string instead, don't worry.

File: Examples/NoModules/Chapter 13/Ex13_12B/Ex13_12B.cpp (hunk 5 above)
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 13/Ex13_12B/Ex13_12B.cpp:1]: The example is there, just not named. I'll add a comment.

File: Exercises/NoModules/Chapter 13/Soln13_05/Box.h (hunk 9 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 13/Soln13_05/Box.h:34]: Yeah, didn't want to everything at once. At least this way they also have one more exercise on implementing a unary operator.

File: Exercises/NoModules/Chapter 13/Soln13_07/Rational.h (hunk 11 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 13/Soln13_07/Rational.h:1]: I did the same. Let's have some faith in few-years-ago Peter.

File: Exercises/NoModules/Chapter 13/Soln13_09/Truckload.cpp (hunk 13 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 13/Soln13_09/Truckload.cpp:50]: That copy-and-swap was a bonus solution. The idiom is not really covered  by the text until Chapter 17, though.

File: Exercises/NoModules/Chapter 13/Soln13_03/Box.cpp (hunk 8 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 13/Soln13_03/Box.cpp:51]: Same as before: exceptions are not known yet. And besides, I prefer this to behave as it does now, precisely like the built-in operator.

**Pull request merged**